app/src/main/kotlin/compiler/Compiler.kt

```diff
@@ -40,17 +54,72 @@ class Compiler(val diagnostics: Diagnostics) {
 
             val programProperties = ProgramAnalyzer.analyzeProgram(astWithBuiltinFunctions, diagnostics)
 
-            val functionCFGs = createGraphForProgram(astWithBuiltinFunctions, programProperties, diagnostics, diagnostics.hasAnyError())
+            val functionDetailsGenerators = FunctionDependenciesAnalyzer.createFunctionDetailsGenerators(
+                astWithBuiltinFunctions,
+                programProperties.variableProperties,
+                programProperties.functionReturnedValueVariables,
+                diagnostics.hasAnyError()
+            )
+
+            val functionCFGs = createGraphForProgram(astWithBuiltinFunctions, programProperties, functionDetailsGenerators, diagnostics)
+
+            val mainFunction = (
+                ast.globals.find { it is Program.Global.FunctionDefinition && it.function.name == "główna" } as Program.Global.FunctionDefinition?
+                )?.function
+            if (mainFunction == null) {
+                diagnostics.report(Diagnostic.MainFunctionNotFound())
+            }
 
             if (diagnostics.hasAnyError())
                 return false
 
-            // TODO:
-            // - make sure "główna" is present and generate a jump to it from "main",
-            // - linearize CFG for each function,
-            // - run liveness analysis on each linear function code,
-            // - run register allocation on each linear function code,
-            // - write ASM to output (display, global variables, functions)
+            val linearFunctions = functionCFGs.mapValues { Linearization.linearize(it.value, covering) }
+
+            val finalCode = linearFunctions.mapValues {
+                Allocation.allocateRegistersWithSpillsHandling(
+                    it.value,
+                    Liveness.computeLiveness(it.value),
+                    Allocation.REGISTER_ORDER,
+                    ColoringAllocation
+                )
+            }
+
+            finalCode.entries.forEach { functionDetailsGenerators[it.key]!!.spilledRegistersOffset.settledValue = it.value.spilledOffset.toLong() }
+
+            output.println("SECTION .bss")
+            DisplayStorage(programProperties.staticDepth).writeAsm(output)
+
+            output.println("\nSECTION .data")
+            GlobalVariableStorage(ast).writeAsm(output)
+
+            output.println("\nSECTION .text")
+            functionDetailsGenerators
+                .filter { it.key.implementation is Function.Implementation.Foreign }
+                .forEach { output.println("extern ${it.value.identifier}") }
+            output.println(
+                """
+                    global main
+                    main:
+                        call ${functionDetailsGenerators[mainFunction]!!.identifier}
+                        ret
```

Comment:
Doing this that way will break the stack alignment. We should either use JMP here or construct valid stack frame (i.e. "push rbp" followed by "mov rbp, rsp").

Reply:
We should also return 0 as exit code. (`mov rax, 0`).

Reply:
We could check here whether `główna` function return type is Unit or Number. If it's Number, then we don't touch RAX and treat the return value as exit code. Otherwise, we should indeed set RAX to zero (why not use _idiomatic_ `xor rax, rax` here).